### ISSUE TITLE
"triton ssh" could support proxying through a bastion host

### ISSUE THREAD

**Issue by jclulow** [joyent/node-triton#259]
Picture if you will, a set of zones with no publicly accessible IP interface.  In order to be able to SSH to those zones, you need some way of passing the SSH connection from a network you can access into the private fabric.  One way to achieve this is to use the `ProxyJump` option (documented in ssh_config(4)_) which accepts another SSH host to use as a proxy for the final target connection.

For example, `ssh -o ProxyJump=bastion interior0` would connect to the `bastion` zone below (solid line), and use that connection as a tunnel to make a connection to the private IP of `interior0` (dotted line).  The same `ProxyJump` option would work in connecting the user to `interior1` as well.

```
     +-----------------+
     |                 |
     | You,            |
     | an intellectual |
     |                 |
     +-----------------+
             :|
             :|
             SSH
             :|
             :|public ip
     +-------:v--------+
     |       :         |
     |     bastion     |   
     |       :         |
     +-------:---------+
             : private ip
             :
  -----------:-----------------------------------
    P R I V A:T E   F A B R I C   N E T W O R K
  -----------:-----------------------------------
             :`................,
             : private ip      : private ip
     +-------v---------+  +----v------------+
     |                 |  |                 |
     |    interior0    |  |    interior1    | . . .
     |                 |  |                 |
     +-----------------+  +-----------------+
```

The `triton ssh` command could read a tag from the target instance; e.g., `tritoncli.ssh.proxy`.  If set, we would _also_ look up the `primaryIp` of the proxy host (`bastion`, above) and pass it to `ssh` via the `ProxyJump` option.  If the special tag is not set on the target, we would continue to connect directly as we do today.

**Comment by jclulow** [comment]
Change available for review @ https://cr.joyent.us/5020

**Comment by jclulow** [comment]
Update to `triton inst ssh --help`:

```
$ triton inst ssh --help
SSH to the primary IP of an instance

Usage:
    triton instance ssh [-h] [USER@]INST [SSH-ARGUMENTS]

Options:
    -h, --help          Show this help.
    --no-proxy          Disable SSH proxy support (ignore "tritoncli.ssh.proxy"
                        tag).

Where INST is the name, id, or short id of an instance. Note that
the INST argument must come before any `ssh` options or arguments.
Where USER is the username to use on the instance. If not specified,
the instance's image is inspected for the default_user tag.
If USER is not specified and the default_user tag is not set, the user
is assumed to be "root".

The "tritoncli.ssh.proxy" tag on the target instance may be set to
the name or the UUID of another instance through which to proxy this
SSH connection.  If set, the primary IP of the proxy instance will be
loaded and passed to SSH via the ProxyJump option.  The --no-proxy
flag can be used to ignore the tag and force a direct connection.

For example, to proxy connections to zone "narnia" through "wardrobe":
    triton instance tag set narnia tritoncli.ssh.proxy=wardrobe

The "tritoncli.ssh.ip" tag on the target instance may be set to the
IP address to use for SSH connections.  This may be useful if the
primary IP address is not available for SSH connections.  This address
must be set to one of the IP addresses attached to the instance.

The "tritoncli.ssh.proxyuser" tag on the proxy instance may be set to
the user account that should be used for the proxy connection (i.e., via
the SSH ProxyJump option).  This is useful when all users of the proxy
instance should use a special common account, and will override the USER
value (if one is provided) for the SSH connection to the target instance.

There is a known issue with SSH connection multiplexing (a.k.a. 
ControlMaster, mux) where stdout/stderr is lost. As a workaround, `ssh`
is spawned with options disabling ControlMaster. See 
<https://github.com/joyent/node-triton/issues/52> for details. If you 
want to use ControlMaster, an alternative is:
    ssh root@$(triton ip INST)
```

**Comment by jclulow** [comment]
### Testing Notes

I've done a lot of manual testing of this change, using all of the different combinations of setting or not setting the tags, with and without the `USER@` portion of the command argument, etc.  I changed the tags and argument to reflect a user that _does_ work, and a user that _doesn't_ work in each case, to make sure that I was correctly hitting the updated configuration.

I also verified with `truss -f -t exec -v exec -a triton ssh ...` that the correct arguments were generated by `triton ssh` for the target SSH connection _and_ for the proxy SSH child process created as part of the use of the `ProxyJump` option to `ssh`.

**Comment by jclulow** [comment]
Put back and published to NPM as `triton@6.3.0`.
